File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.cpp
class Solution {
public:
    // Approach: hashtable
    // record the rows and cols that are to be changed to 0
    // Time: O(mn)

    void setZeroes(vector<vector<int>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();

        unordered_set<int> rows, cols;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] == 0) {
                    rows.insert(i);
                    cols.insert(j);
                }
            }
        }

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (rows.count(i) || cols.count(j)) {
                    matrix[i][j] = 0;
                }
            }
        }
    }
};

File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.py
# class Solution:
#     def setZeroes(self, matrix: List[List[int]]) -> None:
#         """
#         Do not return anything, modify matrix in-place instead.
#         """
#         m, n = len(matrix), len(matrix[0])
        
#         rows = [False] * m
#         cols = [False] * n

#         for i in range(m):
#             for j in range(n):
#                 if matrix[i][j] == 0:
#                     rows[i] = True
#                     cols[j] = True
        
#         for i in range(m):
#             for j in range(n):
#                 if rows[i] or cols[j]:
#                     matrix[i][j] = 0


class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        m, n = len(matrix), len(matrix[0])

        first_row_has_0 = False
        first_col_has_0 = False

        for i in range(m):
            if matrix[i][0] == 0:
                first_col_has_0 = True
                break

        for j in range(n):
            if matrix[0][j] == 0:
                first_row_has_0 = True
                break

        for i in range(1, m):
            for j in range(1, n):
                if matrix[i][j] == 0:
                    matrix[i][0] = matrix[0][j] = 0
        
        for i in range(1, m):
            for j in range(1, n):
                if matrix[i][0] == 0 or matrix[0][j] == 0:
                    matrix[i][j] = 0
        
        if first_col_has_0:
            for i in range(m):
                matrix[i][0] = 0
        if first_row_has_0:
            for j in range(n):
                matrix[0][j] = 0